mesonbuild/modules/simd.py

```diff
@@ -78,35 +78,38 @@ def check(self, state: ModuleState, args: T.Tuple[str], kwargs: CheckKw) -> T.Li
         if 'sources' in kwargs:
             raise mesonlib.MesonException('SIMD module does not support the "sources" keyword')
         prefix = args[0]
-        basic_kwargs = {}
-        for key, value in kwargs.items():
-            if key not in ISETS and key != 'compiler':
-                basic_kwargs[key] = value
+        local_keys = set((*ISETS, 'compiler'))
+        static_lib_kwargs = T.cast('kwtypes.StaticLibrary', {k: v for k, v in kwargs.items() if k not in local_keys})
         compiler = kwargs['compiler']
         conf = build.ConfigurationData()
         for iset in ISETS:
-            if iset not in kwargs:
+            sources = kwargs[iset] # type: ignore
+
+            compile_args = compiler.get_instruction_set_args(iset)
+            if compile_args is None:
+                mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
                 continue
-            iset_fname = kwargs[iset] # Might also be an array or Files. static_library will validate.
-            args = compiler.get_instruction_set_args(iset)
-            if args is None:
-                mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
+
+            if not compiler.has_multi_arguments(compile_args, state.environment)[0]:
+                mlog.log(f'Compiler supports {iset}:', mlog.red('NO'))
                 continue
-            if args:
-                if not compiler.has_multi_arguments(args, state.environment)[0]:
-                    mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
-                    continue
-            mlog.log('Compiler supports %s:' % iset, mlog.green('YES'))
-            conf.values['HAVE_' + iset.upper()] = ('1', 'Compiler supports %s.' % iset)
+            mlog.log(f'Compiler supports {iset}:', mlog.green('YES'))
+            conf.values['HAVE_' + iset.upper()] = ('1', f'Compiler supports {iset}.')
+
             libname = prefix + '_' + iset
-            lib_kwargs = {'sources': iset_fname,
-                          }
-            lib_kwargs.update(basic_kwargs)
+            lib_kwargs = static_lib_kwargs.copy()
+            lib_kwargs['sources'] = sources # type: ignore
+
+            # Add compile args we derived above to those the user provided us
             langarg_key = compiler.get_language() + '_args'
-            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key)
-            all_lang_args = old_lang_args + args
-            lib_kwargs[langarg_key] = all_lang_args
-            result.append(self.interpreter.func_static_lib(None, [libname], lib_kwargs))
+            old_lang_args = mesonlib.extract_as_list(lib_kwargs, langarg_key) # type: ignore
+            all_lang_args = old_lang_args + compile_args
+            lib_kwargs[langarg_key] = all_lang_args # type: ignore
```

Comment:
These type: ignores, can be removed if you instead add `sources: FileOrString` as attributes of `interpreter/kwargs.py -> _BuildTarget`.

Reply:
Does this require any other build typing changes?

Reply:
I made that change and mypy ran cleanly on the entire codebase afterward.

I think at the moment we are mostly relying on interpreter.py not actually getting checked, to hide the fact that most keys in kwargs.py aren't actually filled in. They tend to get filled in while someone is actually working on the corresponding type_checking.py updates, but it's not a requirement I don't think.

Reply:
I will add the sources kwarg in a separate commit. Thank you for pointing that out. I will use a `T.Optional[T.List[FileOrString]]`

Reply:
Should it be on _BaseBuildTarget

Reply:
Thinking again on this:
- the reason it works for simd.py is, more or less, that we are assigning one CheckKw.sources to another, so the type simply does not matter and mypy will be happy since it is the *same*
- you are of course right that this should be a T.List, duh, I do not know what I was thinking
- There's actually several different things it can be as well, in type_checking.py it is defined as: `T.List[T.Union[FileOrString, CustomTarget, CustomTargetIndex, GeneratedList]]` but this is the `declare_dependency(sources:)` definition -- we actually want to support StructuredSources and ExtractedObjects too
  - this is a bug in the PR, I *think*. People could pass ExtractedObjects to a simd.check() and it would currently work because static_library() supports it. Probably no one uses structured_sources here since that implies building a rust library...
- `_BaseBuildTarget` suggests that `jar()` accepts the kwarg. But jar is special since it accepts fewer types, namely, does not accept structured_sources IIRC.

Reply:
That's wrong in several different ways.

1. mypy now believes that `sources` is always present. Its not
2. the type is wrong, it's not `FileOrString`, its `SourceVaragsTypes`
3. this is going to be broken again anyway once the correct annotation is added because lists are invariant

This is tangential to this change. mypy will happily error on "unneeded type ignore", and we should take advantage of that

Reply:
So I should revert?

Reply:
Here is static library: https://mesonbuild.com/Reference-manual_functions.html#static_library.

Reply:
> That's wrong in several different ways.
> 
>  1. mypy now believes that `sources` is always present. Its not

So does this mean CheckKw is badly defined?

>  2. the type is wrong, it's not `T.List[FileOrString]`, its `SourceVaragsTypes`

Yeah, I pointed that out already. Sorry for the initial red herring. :(

>  3. this is going to be broken again anyway once the correct annotation is added because lists are invariant

If it breaks anything it will break it identically now or in the future, so either we *never* handle it correctly or we... cast it? I guess? We should be able to do it today, I think.

Reply:
Regardless of anything else, this conversation made me discover a regression: #12302

Reply:
> So does this mean CheckKw is badly defined?

Yes, whoops.

Reply:
Where is `SourceVaragsTypes`?

Reply:
That typo is better spelled `SourcesVarargsType` and is in type_checking.py

Reply:
Casting Is i think the only way to do that today.

> So does this mean CheckKw is badly defined?

technically yes. but in a not at all easy way to fix. It really should have `sources: NotRequired[SourceVarargsTypes]`, but I think the slightly wrong annotation is preferable to having to redefine everything and keep two copies or even the mad inheritance hacks.

Also, I really dislike the `sources` keyword argument, since it just duplicates positional arguments. This is just another reason to not like it.

Reply:
Ctrl + F returns nothing for me

Reply:
Guess I just needed a rebase...

Reply:
So do I keep my changes to `_BaseBuildTarget` (but switch to SourcesVarargsType), move those changes to `_BuildTarget`, or add the type ignores back and forget it

Reply:
This is now independently implemented so I think the remaining typing issue is to reuse `sse: SourcesVarargsType` etc. in `CheckKw`.